### ISSUE TITLE
filter_var/getenv and the returned value in project.yaml

### ISSUE THREAD

**Issue by jamesmacwhite** [craftcms/cms#4042]
### Description

I'm not sure if this is a bug, but in various configuration files I often use `filter_var()` and `getenv()` in conjunction with each other. Take a custom environment variable of `CRAFT_CACHE_ENABLED` as an example. This is designed to have the value of 0 or 1 in `.env` but the value will be treated as a boolean, so when referenced in a config, I would do:

```php
<?php
return [
    'someConfigName' => filter_var(getenv('CRAFT_CACHE_ENABLED', FILTER_VALIDATE_BOOLEAN))
];
```
0 would be translated to false, 1 being true, rather than having to return the boolean values with the ternary operator or something more expressive.

This works fine, but then I noticed that when doing this with a config parameter that get's written to `project.yaml`, it seems `project.yaml` writes the env var value as a literal string value, not a boolean. So it ends up being:

```yaml
some-handle:
    settings:
      someConfigName: '0'
...
```
Is this the expected behaviour and does it have an impact of how the value is used? I noticed that an alternative expression method appears to work if you use type casting without `filter_var()`:

```php
<?php
return [
    'someConfigName' => (bool) getenv('CRAFT_CACHE_ENABLED')
];
```

What is the recommended approach?

### Steps to reproduce

1. Create an environment variable that acts as a boolean, with a value of 0/1.
2. Reference this environment variable that is written to `project.yaml` using `filter_var()` and `FILTER_VALIDATION_BOOLEAN`
3. Observe the value set in `project.yaml`, it will be a string as 0 or 1, not a boolean value of true/false.

### Additional info

- Craft version: 3.1.19
- PHP version: 7.1
- Database driver & version: MySQL 5.6

**Comment by brandonkelly** [comment]
`filter_var($val, FILTER_VALIDATE_BOOLEAN)` and `(bool)$val` are interchangeable so long as `$val` is always set to `'0'` or `'1'`. Most likely the times you see `'0'` or `'1'` in project.yaml, the value hadn’t actually come from the PHP config file, but the database. Especially if you’re using MySQL, which A) doesn’t support `boolean` column types, and B) only returns data as strings, even if coming from a numeric column, so “boolean” config settings often end up getting represented as `'0'` or `'1'` strings in memory.

> Is this the expected behaviour and does it have an impact of how the value is used?

If using MySQL, yes, and no it won’t have any impact on the application behavior. Once PHP 7.4 is released and we’ve released a Craft update that requires it, we will start moving over to using [typed properties](https://wiki.php.net/rfc/typed_properties_v2) which will solve the problem for good, as those string values will start getting auto-converted to booleans as they are set on their typed model properties, so they’ll be normalized before getting saved out to the project config.

**Comment by jamesmacwhite** [comment]
@brandonkelly Thanks Brandon. I appreciate you taking the time to clarify details on this. I'll continue with my existing methods knowing the the overall application behaviour is not impacted.

Thanks again.

**Comment by brandonkelly** [comment]
We just released Craft 3.1.24 which properly typecasts all core integer and boolean values as they’re saved to the project config (#4167). Up to plugin developers to follow suit, but this should take care of most of the awkwardness.

**Comment by jamesmacwhite** [comment]
Great! Thanks Brandon!
